File: solution/java/week02-object-calisthenics/exercise2/src/test/java/BookStoreTest.java
import org.junit.Test;
import store.Book;
import store.BookStore;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class BookStoreTest {

    private final BookStoreForTest store = new BookStoreForTest();

    private static Book createABook(int copies) {
        return Book.tryCreateBook(
                "Lord of the ring",
                "JR Tolkien", copies
        ).orElseThrow();
    }

    private Book addBookToStore(int initialCopies) {
        var createdBook = createABook(initialCopies);
        store.addBook(createdBook.getTitle(), createdBook.getAuthor(), initialCopies);
        return createdBook;
    }

    @Test
    public void addANewBookToStore() {
        var bookToAdd = createABook(2);

        store.addBook(bookToAdd.getTitle(),
                bookToAdd.getAuthor(),
                bookToAdd.getCopies());

        assertTrue("The book should be in the inventory",
                store.verifyBookIsInStore(bookToAdd));
    }

    @Test
    public void doesNotAddBookToStoreIfNoCopy() {
        var noCopy = 0;
        var bookToAdd = createABook(noCopy);

        store.addBook(bookToAdd.getTitle(),
                bookToAdd.getAuthor(),
                noCopy
        );

        assertFalse("The book should not be in the inventory",
                store.verifyBookIsInStore(bookToAdd));
    }

    @Test
    public void doesNotAddBookToStoreIfInvalidTitle() {
        String invalidTitle = null;
        String author = "JR Tolkien";

        store.addBook(invalidTitle, author, 2);

        assertFalse("The book should not be in the inventory",
                store.bookInInventory(invalidTitle, author));
    }

    @Test
    public void doesNotAddBookToStoreIfInvalidAuthor() {
        String title = "Lord of the ring";
        String invalidAuthor = null;

        store.addBook(title, invalidAuthor, 2);

        assertFalse(
                "The book should not be in the inventory",
                store.bookInInventory(title, invalidAuthor)
        );
    }

    @Test
    public void addAnExistingBookToStoreIncreasesTheCopies() {
        var initialCopies = 2;
        var additionalCopies = 3;
        var bookAdded = addBookToStore(initialCopies);
        int expectedNumberOfCopy = additionalCopies + bookAdded.getCopies();

        store.addBook(bookAdded.getTitle(),
                bookAdded.getAuthor(),
                additionalCopies);

        assertTrue(
                "The book should have the following number of copies " + expectedNumberOfCopy,
                store.verifyNumberOfCopyForBook(bookAdded, expectedNumberOfCopy)
        );
    }

    @Test
    public void sellABookFromTheStore() {
        var copiesSold = 1;
        var initialCopies = 2;
        var bookToSell = addBookToStore(initialCopies);
        int expectedNumberOfCopy = bookToSell.getCopies() - copiesSold;

        store.sellBook(bookToSell.getTitle(),
                bookToSell.getAuthor(),
                copiesSold);

        assertTrue(
                "The book should have the following number of copies " + expectedNumberOfCopy,
                store.verifyNumberOfCopyForBook(bookToSell, expectedNumberOfCopy)
        );
    }

    @Test
    public void cannotSellZeroCopyOfABook() {
        var copiesSold = 0;
        var initialCopies = 2;
        var bookToSell = addBookToStore(initialCopies);
        int expectedNumberOfCopy = bookToSell.getCopies();

        store.sellBook(bookToSell.getTitle(),
                bookToSell.getAuthor(),
                copiesSold);

        assertTrue(
                "The book should have the following number of copies " + expectedNumberOfCopy,
                store.verifyNumberOfCopyForBook(bookToSell, expectedNumberOfCopy)
        );
    }

    @Test
    public void sellTheLastCopyOfABookFromTheStore() {
        var copiesSold = 1;
        var bookToSell = addBookToStore(copiesSold);

        store.sellBook(bookToSell.getTitle(),
                bookToSell.getAuthor(),
                copiesSold);

        assertFalse(
                "The book should no longer be in the inventory",
                store.verifyBookIsInStore(bookToSell)
        );
    }

    static class BookStoreForTest extends BookStore {
        public boolean bookInInventory(String title, String author) {
            return getBookBy(title, author).isPresent();
        }

        public boolean bookInInventory(Book bookSearched) {
            return getBookBy(bookSearched.getTitle(), bookSearched.getAuthor()).isPresent();
        }

        public boolean verifyBookIsInStore(Book book) {
            return bookInInventory(book);
        }

        public boolean verifyNumberOfCopyForBook(Book bookAdded, int expectedNumberOfCopy) {
            return verifyBookIsInStore(bookAdded)
                    && getBookBy(bookAdded.getTitle(), bookAdded.getAuthor())
                    .orElseThrow()
                    .getCopies() == expectedNumberOfCopy;

        }
    }
}

File: solution/c#/week02-object-calisthenics/BookStore.Tests/BookStoreTest.cs
using Xunit;

namespace BookStore.Tests
{
    public class BookStoreTest
    {
        private readonly BookStoreUnderTest _store = new();

        [Fact]
        public void AddANewBookToStore()
        {
            var bookToAdd = CreateABook(2);
            _store.AddBook(bookToAdd!.Title, bookToAdd.Author, bookToAdd.Copies);

            Assert.True(_store.VerifyBookIsInStore(bookToAdd), "The book should be in the inventory");
        }


        [Fact]
        public void DoesNotAddBookToStoreIfNoCopy()
        {
            const int noCopy = 0;
            var bookToAdd = CreateABook(noCopy);

            _store.AddBook(bookToAdd!.Title, bookToAdd.Author, noCopy);

            Assert.False(_store.VerifyBookIsInStore(bookToAdd), "The book should not be in the inventory");
        }

        [Fact]
        public void DoesNotAddBookToStoreIfInvalidTitle()
        {
            string? invalidTitle = null;
            const string author = "JR Tolkien";

            _store.AddBook(invalidTitle, author, 2);

            Assert.False(_store.BookInInventory(invalidTitle, author), "The book should not be in the inventory");
        }

        [Fact]
        public void DoesNotAddBookToStoreIfInvalidAuthor()
        {
            const string title = "Lord of the ring";
            string? invalidAuthor = null;

            _store.AddBook(title, invalidAuthor, 2);

            Assert.False(_store.BookInInventory(title, invalidAuthor), "The book should not be in the inventory");
        }

        [Fact]
        public void AddAnExistingBookToStoreIncreasesTheCopies()
        {
            const int initialCopies = 2;
            const int additionalCopies = 3;
            var bookAdded = AddBookToStore(initialCopies);
            var expectedNumberOfCopy = additionalCopies + bookAdded.Copies;

            _store.AddBook(bookAdded.Title, bookAdded.Author, additionalCopies);

            Assert.True(_store.VerifyNumberOfCopyForBook(bookAdded, expectedNumberOfCopy),
                $"The book should have the following number of copies {expectedNumberOfCopy}");
        }

        [Fact]
        public void SellABookFromTheStore()
        {
            const int copiesSold = 1;
            const int initialCopies = 2;
            var bookToSell = AddBookToStore(initialCopies);
            var expectedNumberOfCopy = bookToSell.Copies - copiesSold;

            _store.SellBook(bookToSell.Title, bookToSell.Author, copiesSold);

            Assert.True(_store.VerifyNumberOfCopyForBook(bookToSell, expectedNumberOfCopy),
                $"The book should have the following number of copies {expectedNumberOfCopy}");
        }

        [Fact]
        public void CannotSellZeroCopyOfABook()
        {
            const int copiesSold = 0;
            const int initialCopies = 2;
            var bookToSell = AddBookToStore(initialCopies);
            var expectedNumberOfCopy = bookToSell.Copies;

            _store.SellBook(bookToSell.Title, bookToSell.Author, copiesSold);

            Assert.True(_store.VerifyNumberOfCopyForBook(bookToSell, expectedNumberOfCopy),
                $"The book should have the following number of copies {expectedNumberOfCopy}");
        }

        [Fact]
        public void SellTheLastCopyOfABookFromTheStore()
        {
            const int copiesSold = 1;
            var bookToSell = AddBookToStore(copiesSold);

            _store.SellBook(bookToSell.Title, bookToSell.Author, copiesSold);

            Assert.False(_store.VerifyBookIsInStore(bookToSell), "The book should no longer be in the inventory");
        }

        private static Book? CreateABook(int copies)
            => Book.TryCreateBook("Lord of the ring", "JR Tolkien", copies);

        private Book AddBookToStore(int initialCopies)
        {
            var createdBook = CreateABook(initialCopies);
            _store.AddBook(createdBook!.Title, createdBook.Author, initialCopies);
            return createdBook;
        }

        private class BookStoreUnderTest : BookStore
        {
            public bool BookInInventory(string title, string author) => GetBookBy(title, author) != null;

            private bool BookInInventory(Book bookSearched)
                => GetBookBy(bookSearched.Title, bookSearched.Author) != null;

            public bool VerifyBookIsInStore(Book? book) => book != null && BookInInventory(book);

            public bool VerifyNumberOfCopyForBook(Book? bookAdded, int expectedNumberOfCopy)
                => VerifyBookIsInStore(bookAdded) &&
                   GetBookBy(bookAdded!.Title, bookAdded.Author)!.Copies == expectedNumberOfCopy;
        }
    }
}